modules/stochastic_tools/src/samplers/Metropolis.C

```diff
@@ -0,0 +1,92 @@
+  params.addRequiredParam<dof_id_type>("num_rows", "The number of Markov chains.");
+  params.addRequiredParam<std::vector<DistributionName>>("distributions",
+                                                         "The distribution names to be sampled.");
+  params.addRequiredParam<VectorPostprocessorName>(
+      "inputs_vpp", "Vectorpostprocessor with the samples created by sampler.");
+  params.addRequiredParam<std::vector<std::string>>(
+      "inputs_names",
+      "Name of vector from vectorpostprocessor with the samples created by sampler.");
+  params.addRequiredParam<std::vector<Real>>("proposal_std",
+                                             "Standard deviations of the proposal distributions.");
+  params.addRequiredParam<std::vector<Real>>(
+      "initial_values", "Seed input values to get the Metropolis sampler started.");
+  return params;
+}
+
+Metropolis::Metropolis(const InputParameters & parameters)
+  : Sampler(parameters),
+    VectorPostprocessorInterface(this),
+    _inputs_names(getParam<std::vector<std::string>>("inputs_names")),
+    _proposal_std(getParam<std::vector<Real>>("proposal_std")),
+    _initial_values(getParam<std::vector<Real>>("initial_values")),
+    _step(getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")->timeStep()),
+    _perf_compute_sample(registerTimedSection("computeSample", 4))
+{
+  for (const DistributionName & name : getParam<std::vector<DistributionName>>("distributions"))
+    _distributions.push_back(&getDistributionByName(name));
+
+  if (_distributions.size() != _inputs_names.size() ||
+      _distributions.size() != _proposal_std.size() ||
+      _distributions.size() != _initial_values.size())
+    paramError("distributions",
+               "The number of distributions, input names, proposal standard deviations, and "
+               "initial values should be equal");
+
+  setNumberOfRows(getParam<dof_id_type>("num_rows"));
+  setNumberOfCols(_distributions.size());
+}
+
+void
+Metropolis::sampleSetUp()
+{
+  _values_distributed = isVectorPostprocessorDistributed("inputs_vpp");
+
+  _inputs_ptr.clear();
+  _inputs_ptr.reserve(_inputs_names.size());
+  for (const auto & name : _inputs_names)
+    _inputs_ptr.push_back(&getVectorPostprocessorValue("inputs_vpp", name, !_values_distributed));
+}
+
+Real
+Metropolis::computeSample(dof_id_type row_index, dof_id_type col_index)
+{
+  Real _proposed_sample = 0.0;
+  Real _acceptance_ratio = 0.0;
+
+  TIME_SECTION(_perf_compute_sample);
+  dof_id_type offset = _values_distributed ? getLocalRowBegin() : 0;
+  if (_step > 0)
+  {
+    _proposed_sample = Normal::quantile(
+        getRand(), (*_inputs_ptr[col_index])[row_index - offset], _proposal_std[col_index]);
+    _acceptance_ratio =
+        std::log(_distributions[col_index]->pdf(_proposed_sample)) -
+        std::log(_distributions[col_index]->pdf((*_inputs_ptr[col_index])[row_index - offset]));
+    if (_acceptance_ratio > std::log(getRand()))
```

Comment:
I just noticed you are calling `getRand()` twice here. Are you expecting that the two numbers produced are the same? This could possibly be why you are getting different numbers in parallel.

Reply:
@zachmprince The `getRand()` should be called twice as per the algorithm. First time to propose a new sample and second time to check whether the acceptance ratio is greater than a random number b/w 0 and 1.

I modified the code per our discussion and did not use the vectorpostprocessor interface. Now, the sampler is working well on multiple processor. But the results on single and multiple processors aren't the same. So, in tests, I set `min_parallel` and `max_parallel` to 4.